File: backend/app/models/expense.py
"""
Expense Model
"""
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Enum as SQLEnum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.core.database import Base

class ExpenseStatus(str, enum.Enum):
    DRAFT = "draft"
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"

class Expense(Base):
    __tablename__ = "expenses"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    category_id = Column(Integer, ForeignKey("categories.id"), nullable=False)
    report_id = Column(Integer, ForeignKey("reports.id"), nullable=True)
    trip_id = Column(Integer, ForeignKey("trips.id"), nullable=True)
    
    # Datos del gasto
    amount = Column(Integer, nullable=False)  # En centavos para precisión
    currency = Column(String(3), default="USD", nullable=False)
    merchant = Column(String(255), nullable=True)  # Comercio/Proveedor
    description = Column(Text, nullable=True)
    expense_date = Column(DateTime, nullable=False)
    
    # OCR y recibo
    receipt_url = Column(String(500), nullable=True)
    receipt_original_name = Column(String(255), nullable=True)
    ocr_data = Column(Text, nullable=True)  # JSON con datos extraídos
    ocr_confidence = Column(Integer, nullable=True)  # 0-100
    
    # Estado
    status = Column(SQLEnum(ExpenseStatus), default=ExpenseStatus.DRAFT, nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    user = relationship("User", back_populates="expenses")
    category = relationship("Category", back_populates="expenses")
    report = relationship("Report", back_populates="expenses")
    trip = relationship("Trip", back_populates="expenses")

File: mobile/lib/models/expense.dart
import 'category.dart';

class Expense {
  final int id;
  final int userId;
  final int categoryId;
  final int amount; // En centavos
  final String currency;
  final String? merchant;
  final String? description;
  final DateTime expenseDate;
  final String status;
  final int? reportId;
  final String? receiptUrl;
  final DateTime createdAt;
  final Category? category;

  Expense({
    required this.id,
    required this.userId,
    required this.categoryId,
    required this.amount,
    required this.currency,
    this.merchant,
    this.description,
    required this.expenseDate,
    required this.status,
    this.reportId,
    this.receiptUrl,
    required this.createdAt,
    this.category,
  });

  factory Expense.fromJson(Map<String, dynamic> json) {
    return Expense(
      id: json['id'],
      userId: json['user_id'],
      categoryId: json['category_id'],
      amount: json['amount'],
      currency: json['currency'] ?? 'USD',
      merchant: json['merchant'],
      description: json['description'],
      expenseDate: DateTime.parse(json['expense_date']),
      status: json['status'],
      reportId: json['report_id'],
      receiptUrl: json['receipt_url'],
      createdAt: DateTime.parse(json['created_at']),
      category: json['category'] != null 
          ? Category.fromJson(json['category']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category_id': categoryId,
      'amount': amount,
      'currency': currency,
      'merchant': merchant,
      'description': description,
      'expense_date': expenseDate.toIso8601String(),
    };
  }

  double get amountInDollars => amount / 100.0;
}
